numbat/src/parser.rs

```diff
@@ -392,447 +392,493 @@ impl<'a> Parser<'a> {
-                        kind: ParseErrorKind::AliasUsedOnFunction,
+                        kind: ParseErrorKind::ExpectedParameterNameInFunctionDefinition,
                         span: self.peek().span,
                     });
                 }
+            }
 
-                let mut decorators = vec![];
-                std::mem::swap(&mut decorators, &mut self.decorator_stack);
+            let return_type_annotation = if self.match_exact(TokenKind::Arrow).is_some() {
+                Some(self.type_annotation()?)
+            } else {
+                None
+            };
 
-                Ok(Statement::DefineFunction {
-                    function_name_span,
-                    function_name: fn_name.lexeme.clone(),
-                    type_parameters,
-                    parameters,
-                    body,
-                    return_type_annotation,
-                    decorators,
-                })
+            let (body, local_variables) = if self.match_exact(TokenKind::Equal).is_none() {
+                (None, vec![])
             } else {
-                Err(ParseError {
-                    kind: ParseErrorKind::ExpectedIdentifierAfterFn,
-                    span: self.peek().span,
-                })
-            }
-        } else if self.match_exact(TokenKind::Dimension).is_some() {
-            if let Some(identifier) = self.match_exact(TokenKind::Identifier) {
-                if identifier.lexeme.starts_with("__") {
-                    return Err(ParseError::new(
-                        ParseErrorKind::DoubleUnderscoreTypeNamesReserved,
-                        identifier.span,
-                    ));
-                }
+                self.skip_empty_lines();
+                let body = self.expression()?;
 
-                if self.match_exact(TokenKind::Equal).is_some() {
-                    self.skip_empty_lines();
-                    let mut dexprs = vec![self.dimension_expression()?];
+                // After parsing the `where` statement we have eaten all
+                // the newlines. We're not supposed to do that.
+                // We'll resume the parser to where we were at before doing that.
+                let mut step_back_to = self.current;
 
-                    while self.match_exact(TokenKind::Equal).is_some() {
+                let mut local_variables = Vec::new();
+                self.skip_empty_lines();
+                if self.match_exact(TokenKind::Where).is_some() {
+                    step_back_to = self.current;
+                    self.skip_empty_lines();
+                    while let Ok(local_variable) = self.parse_variable() {
+                        local_variables.push(local_variable);
+                        step_back_to = self.current;
                         self.skip_empty_lines();
-                        dexprs.push(self.dimension_expression()?);
                     }
+                }
+                self.current = step_back_to;
```

Comment:
This might be the problem with the missing decorators. `self.current` is not the only stateful thing in the parser. There's also the `decorator_stack`, which would need to be reset.

Reply:
This was not the problem, but I fixed that.